#169610

nan-tic-dev opened the pull request.

dCeylan committed:
Added tree view
task #169610

KevinNantic committed:
Remade module, free of charge option on purchases
Added new boolean field to purchase line. When active, unit_price is set to 0.
However, the move generated will have its product's cost price as unit price, to avoid altering average cost price.
Once again, when advancing the move to state "done", unit price is set to product's unit price.
Additionally, purchase lines with this mark will NOT be taken into account as invoice line on the purchase's invoice.

Task #169610

albertca reviewed (commented):
Falta un scenario

albertca commented on purchase.py:
```diff
@@ -0,0 +1,27 @@
+class PurchaseLine(metaclass=PoolMeta):
+    __name__ = 'purchase.line'
+
+    charge_free = fields.Boolean('Free of Charge', states={
```
Naming the field charge_free is inconsistent with the name of the module itself and the label.

Why not naming it free_of_charge?

albertca commented on stock.py:
```diff
@@ -0,0 +1,14 @@
+        PurchaseLine = Pool().get('purchase.line')
+        for move in moves:
+            if move.origin and isinstance(move.origin, PurchaseLine):
+                if move.origin.charge_free and move.product:
```
No need to ensure `move.product` evaluates tu True given that the `product` field in `stock.move` is required.

albertca commented on view/sale_view_tree.xml:
```diff
@@ -0,0 +1,5 @@
+    <xpath expr="/tree/field[@name='amount']" position="after">
+        <field name="free"/>
+    </xpath>
+</data>
```
This file must be removed. The tests probably failed.

KevinNantic committed:
Code Review corrections
Added test

Task #169610

KevinNantic committed:
Removed unnecessary default test

Task #169610

albertca commented on tests/test_module.py:
```diff
@@ -1,11 +0,0 @@
-# This file is part purchase_line_free_of_charge module for Tryton.
```
Aquest fitxer no s'ha d'eliminar! T'has petat els tests estàndard de mòdul!

albertca commented on tests/test_purchase_free_of_charge.py:
```diff
@@ -0,0 +1,143 @@
+        ProductUom = Model.get('product.uom')
+        unit, = ProductUom.find([('name', '=', 'Unit')])
+        ProductTemplate = Model.get('product.template')
+        template = ProductTemplate()
+        template.name = 'product'
+        template.default_uom = unit
+        template.type = 'goods'
+        template.purchasable = True
+        template.list_price = Decimal('10')
+        template.account_category = account_category
+        template.save()
```
cost_method should be set to 'average' and when the goods are received ensure that the product has the correct cost_price.

The reason is that we may set the unit_price of the move correctly but at the wrong time and the value may not be updated appropriately.

KevinNantic committed:
Restored default test/improved test scenario
Restored mistakenly removed default test
Added onto test scenario, so now checks cost price at every step of the shipment receivement,
including reciving another shipment that modifies cost price inbetween creating and receiving the original shipment.

Task #169610

KevinNantic committed:
Improved test/solved bug

Updated test to better represent the modules use case.
Fixed bug where a move's unit_price was not being properly updated to product's cost price
when the purchase it came from was free of charge.

Task #169610
